### PR TITLE
 [codegen][debug info] generating inline function information aka C 'inline' keyword and more

### PR DESCRIPTION
**Pull request by vvlevchenko**

### CODE REVIEW

**Comment by olonho** [comment]
Neep to update compiler for that to work?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt [hunk 1]
```diff
@@ -22,6 +22,7 @@ import org.jetbrains.kotlin.backend.konan.ir.*
 internal class CodeGenerator(override val context: Context) : ContextUtils {
 
     fun llvmFunction(function: IrFunction): LLVMValueRef = function.llvmFunction
+    fun llvmFunctionOrNull(function: IrFunction): LLVMValueRef? = function.llvmFunctionOrNull
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt]: Shalln't one be expressed via another?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/DebugUtils.kt [hunk 2]
```diff
@@ -63,6 +63,8 @@ fun KonanConfig.debugInfoVersion():Int = configuration[KonanConfigKeys.DEBUG_INF
 internal class DebugInfo internal constructor(override val context: Context):ContextUtils {
     val files = mutableMapOf<String, DIFileRef>()
     val subprograms = mutableMapOf<LLVMValueRef, DISubprogramRef>()
+    /* Some functions inlined out and haven't got LLVM value. */
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/DebugUtils.kt]: "Inlined out" :). Maybe something like
"Some functions are inlined on all callsites and body is eliminated by DCE, so there's no LLVM value"?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 3]
```diff
@@ -1619,6 +1619,15 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
 
         override fun returnableBlockScope(): CodeContext? = this
 
+        override fun location(line:Int, column: Int): LocationInfo? {
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]:
```line: Int
       ^
```

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/LlvmDeclarations.kt [hunk 4]
```diff
@@ -41,8 +41,7 @@ internal class LlvmDeclarations(
     private val fields: Map<IrField, FieldLlvmDeclarations>,
     private val staticFields: Map<IrField, StaticFieldLlvmDeclarations>,
     private val unique: Map<UniqueKind, UniqueLlvmDeclarations>) {
-    fun forFunction(function: IrFunction) = functions[function] ?:
-            error("${function.descriptor} in ${function.parent.fqNameSafe}")
+    fun forFunction(function: IrFunction) = functions[function]
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/LlvmDeclarations.kt]: So here signature is silently changed. Maybe add another one? Or smth like
```
fun forFunction(function: IrFunction, orElse: (IrFunction) -> Unit = {
  it -> error("${it.descriptor} in ${it.parent.fqNameSafe}")}) = 
  functions[function] ?: orElse(function)
```

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/LLVMCoverageWriter.kt [hunk 5]
```diff
@@ -56,7 +56,7 @@ internal class LLVMCoverageWriter(
                         fileIds.toCValues(), fileIds.size.signExtend(),
                         regions.toCValues(), regions.size.signExtend())
 
-                val functionName = context.llvmDeclarations.forFunction(functionRegions.function).llvmFunction.name
+                val functionName = context.llvmDeclarations.forFunction(functionRegions.function)!!.llvmFunction.name
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/LLVMCoverageWriter.kt]: Not sure if NPE is better than diagnostics we had before.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/LlvmDeclarations.kt (hunk 4 above)
Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/LlvmDeclarations.kt]: Then signature will change even more dramatically. But I understand idea and will try to develop to something useful

**Pull request merged**